### DISCUSSION TITLE
"Create" sensor

### DISCUSSION THREAD

**Discussion by zmechu** [arendst/Tasmota#18856] [Q&A] [answered]
Hello!
I'm using one pf Tasmota I/O pin as pulse counter. Then sends counter data to Home Assistant via MQTT. Under HA use Helper to add counter unit of measurements and proper value (divide by constant).

But... is it possible to handle it with Tasmota?

My questions:
1. is there a way to add to Tasmota's Counter unit of measurements? For example "Counter1 2798321" into " "Counter1 2798321 mA" or sth like that?
2. Is there a way to "create" kind of "virtual sensor"? I'd like to "create" new sensor (for example "Current") and using "Counter1" value inform Home Assistant about new entity under Tasmota device.

Thanks for help!

**Comment by sfromis** [comment]
Without modifying the project source code, you cannot add units to the sensor payloads. Tasmota already does not add units, except for the very few values where there are existing options for selecting what unit you want, like C/F. For other values, the unit is implicit, and expected to be understood by the subscriber. Thus, the units the driver may include on the web UI are not passed to the JSON. What happens with most units in HA is that the Tasmota integration has setup for "knowing" what JSON keys are in what units, as this is not passed from Tasmota.

With Tasmota on ESP32, you do have the Berry language, including easily accessible features for coding a sensor driver to add new sensors and values, to both the JSON and Tasmota front page, working like the sensors already part of the project. I'd expect such sensors to also be recognized by HA, but there is still no format for passing units. While you could include a field with a unit in the JSON payload, it would not be automatically recognized as such, due to Tasmota not having a standard for doing that.

On ESP8266, I think that you can do something similar with the scripting feature, but I've not been using that.

**Comment by zmechu** [comment]
Thanks for answer!
I did trick, but doesn't work (I don't know why) - instead of MQTT message:
` tele/tasmota_CB6F4E/SENSOR = {"Time":"2023-06-09T13:21:24","COUNTER":{"C1":1}}`
make rule:
`[...]DO Publish tele/%topic%/SENSOR {"Time":"%timestamp%","COUNTER":{"C1":%value%},"CURRENT":{"mA":%value%}} ENDON`
HA didn't "notice" new data as entity or can't find how to use this data.

On ESP8266 there is https://tasmota.github.io/docs/Smart-Meter-Interface/ but till now I'm not so skillful to use it :-)

**Reply by sfromis** [reply]
As HA automatic discovery is based on messages Tasmota publishes when restarting, your additional publish payloads are too late for that. You can still do manual configuration to pick up such.

**Reply by zmechu** [reply]
Is there a way to add information about my additional payload when restarting?

**Reply by sfromis** [reply]
No, the values automatically recognized has to be in the first "sample" payload published to `tasmota/discovery/<mac>/sensors`, which you can get if you create a driver in Berry on ESP32. I think that it may be doable to instead publish messages the more complex format for the older MQTT autodiscovery feature no longer included in Tasmota builds. Simplest is probably to just do the manual config in HA.

**Reply by zmechu** [reply, marked as answer]
Seems to works using scripts :-)

`
>D
>B
=>sensor53 r
=>sensor53 c1
>M 1
+1,12,c,0,-10,METER
1,1-0:1.8.0*255(@1,current,mA,current,16
#
`



It creates one "virtual" sensor - in my system a counter on GPIO12, which add single mA into counter value and returns MQTT:

`MQT: tele/tasmota_CB6F4E/SENSOR = {"Time":"2023-06-12T18:55:20","METER":{"current":2835262}}`

Only one thing - there is no "mA" unit, but it is 95% what I need.

Edit1: unit of measurement is Home Assistant related, not Tasmota! To properly recognize unit, sensor must be one of Device Class: https://www.home-assistant.io/integrations/sensor/ - then sensor class AND unit will be properly decoded. Voila!

Edit2: it is possible to create own unit of measurement. Just customize your entity in HA configuration.yaml:
`
  customize:
    sensor.your_sensor_entity_name:
      unit_of_measurement : abcdef
`

**Comment by nikito7** [comment]
Old HA discovery, if anyone need it:

```
Publish2 homeassistant/sensor/nikito7-EB1/signal/config {"name":"ESP Signal","uniq_id":"EB1_ESP_Signal","stat_t":"tele/edpbox1/STATE","val_tpl":"{{ value_json.Wifi.Signal }}","unit_of_meas":"dB","dev_cla":"signal_strength","dev":{"ids":"nikito7-EB1"}}
```
